design/cassandra/design.md

```diff
@@ -0,0 +1,112 @@
+# Kubernetes Operator for Apache Cassandra
+
+## Why ?
+
+* Cassandra is one of the most popular NoSQL databases.
+* Peer-to-Peer write-anywhere design for very high write and read throughput.
+* Despite those advantages, maintaining a Cassandra cluster is a highly manual and tiring task, which requires extensive familiarity with Cassandra's internals.
+* An operator will automate all the manual interventions that a human administrator normally has to perform and provide a production-ready database out-of-the-box.
+
+## Features
+
+The operator will provide all the important features of Cassandra while exposing a user-friendly declarative UI, in accordance to Kubernetes principles.
+More specifically, those features include:
+
+1. Cluster Creation
+2. (Auto)Scaling 
+3. Failed Cassandra Node Replacement
+4. Schedule Node Repair
+5. Backup/Restore
+6. Metrics endpoint for Kubernetes
+
```

Comment:
Are there other Cassandra operators out there that you can list here?

Reply:
The main ones I know of are:

* [cassandra-operator](https://github.com/instaclustr/cassandra-operator) by [Instaclustr](https://www.instaclustr.com/): This is indeed a very promising project, coming from a company that offers managed Cassandra deployments. It also utilizes a sidecar, which accepts commands through HTTP endpoints. It is written in Java, so it misses on the more advanced functionality of client-go. Also I don't know how many people are out there developing Kubernetes operators in other than Golang languages, so that may be a barrier for the community the operator will attract. They also do not support different racks.
  * *Phase:* Alpha
  * *License:* Apache 2.0

* [navigator](https://github.com/jetstack/navigator) by [Jetstack](https://www.jetstack.io/): Navigator is aspiring to be a managed DBaaS platform on top of Kubernetes. Their model they promote is very similar to the one I am proposing (CRD + Controller + Sidecar) and has been an inspiration while designing the model for Cassandra Operator. They currently provide operators for Cassandra and Elasticsearch.
  * *Phase:* Alpha
  * *License:* Apache 2.0

There are also a couple of Helm charts out there that try to run Cassandra using StatefulSets only, utilizing things like lifecycle hooks with custom scripts. While they certainly can work, they lack the amount of control that an operator provides. 

Major Point of Differentiation:
* *Static IP:* While StatefulSets provide a stable identity through an FQDN, this is unfortunately not enough for Cassandra, as it identifies its instances by their IP address. Starting a Pod without providing a thing that will serve as the Pod's unique ID makes you subject to race conditions in edge case scenarios. 
  * *Example:* A Cassandra pod starts and bootstraps, but the node fails before you manage to get the node identity (UUID) from the JMX interface. There is no way to know what that instance's ID was, so you can find its IP from gossip to replace it. You can maybe look for dead instances and replace them, but what if there are multiple of them ? This will require manual intervention. By having a Static IP on each Cassandra Pod, we simplify the process of replacing an instance while avoiding race conditions by design, because *we* provide the instance's identity.

I also believe this operator will provide the rook framework with a new pattern (CRD + Controller + Sidecar) that other storage providers can leverage to build their operators.

Reply:
For what it's worth right now, I'm working on setting up a Scylla cluster on Kubernetes, though with a more manual approach with stateful sets. The last thing I was working on was resolving various networking issues. You are pretty much spot on in terms of the issues that happen due to ephemeral IPs due to the pod's nature. 
Couple things I currently have set up to work around this. Pod and Host IPs resolve differently and sometimes create issues while gossiping due to the dual IPs. Ie if a pod dies and gets restarted, on join it broadcasts with the host IP instead of pod IP. After a long enough period, all of them have converged to their host IPs. On the next pod restart, it fails to figure out the seeds and cluster info thus starts failing. The resolution is to use host networking so the pod binds to the network stack natively and unifies the address. However, this introduces a bunch of other issues. One of the main ones being that with host networking you lose the DNS resolver from Kubernetes thus can no longer resolve nodes by its hostname which brings us to the issue of changing IPs with no mechanism to fixate them. The real solution to the issue is currently pending a PR (https://github.com/kubernetes/kubernetes/pull/68293) which transfers the responsibility to K8s to manage the DNS for the hostname in regards to the changing IPs.
A temporary workaround is to use static IPs per node and then label the nodes and use a selector to force individual instances (at least the seed nodes) on those known nodes. 
In the meanwhile, host networking is at least a bit more stable as full node swaps and restarts are rare (at least in regard to pods) and thus less prone to the issue but are still susceptible unless you hack around it which almost beats the purpose of using orchestration at all. Anyways, once the PR is merged into a stable release it will be possible to manage the entire network setup through K8s without hardcoded bindings and dependencies.

Feel free to reach out for more info as I intend to stick around thanks to the awesome guys at Scylla.